src/ai/backend/manager/repositories/permission_controller/db_source/db_source.py

```diff
@@ -80,20 +252,204 @@ async def _get_role(self, db_session: SASession, role_id: uuid.UUID) -> RoleRow:
+        purger: Purger[PermissionGroupRow],
+    ) -> None:
+        """Remove a permission group from a role (private, within transaction).
+
+        Also deletes all permissions in this group.
+        """
+        # Delete all permissions in this group
+        stmt_perms = sa.delete(PermissionRow).where(
+            PermissionRow.permission_group_id == purger.pk_value
+        )
+        await db_session.execute(stmt_perms)
+
+        # Delete the permission group using purger
+        await execute_purger(db_session, purger)
+
+    async def _add_permission_to_group(
+        self,
+        db_session: SASession,
+        creator: Creator[PermissionRow],
+    ) -> PermissionRow:
+        """Add a permission to a permission group (private, within transaction)."""
+        result = await execute_creator(db_session, creator)
+        return result.row
+
+    async def _remove_permission_from_group(
+        self,
+        db_session: SASession,
+        purger: Purger[PermissionRow],
+    ) -> None:
+        """Remove a permission from a permission group (private, within transaction)."""
+        await execute_purger(db_session, purger)
+
+    async def _add_object_permission_to_role(
+        self,
+        db_session: SASession,
+        creator: Creator[ObjectPermissionRow],
+    ) -> ObjectPermissionRow:
+        """Add an object permission to a role (private, within transaction)."""
+        result = await execute_creator(db_session, creator)
+        return result.row
+
+    async def _add_permission_group(
+        self,
+        db_session: SASession,
+        creator: Creator[PermissionGroupRow],
+    ) -> PermissionGroupRow:
+        """Add a permission group to a role (private, within transaction)."""
+        result = await execute_creator(db_session, creator)
+        return result.row
+
+    async def _remove_object_permission_from_role(
+        self,
+        db_session: SASession,
+        purger: Purger[ObjectPermissionRow],
+    ) -> None:
+        """Remove an object permission from a role (private, within transaction)."""
+        await execute_purger(db_session, purger)
+
     async def update_role(self, updater: Updater[RoleRow]) -> RoleRow:
         async with self._db.begin_session() as db_session:
             result = await execute_updater(db_session, updater)
             if result is None:
                 raise ObjectNotFound(f"Role with ID {updater.pk_value} does not exist.")
             return result.row
 
-    async def delete_role(self, data: RoleDeleteInput) -> RoleRow:
+    async def delete_role(self, purger: Purger[RoleRow]) -> RoleRow | None:
+        """
+        Delete a role (soft delete by setting status to DELETED).
+
+        Uses an Updater internally to set the status field.
+        """
+        from ....types import OptionalState
+
+        spec = RoleUpdaterSpec(status=OptionalState.update(RoleStatus.DELETED))
+        updater = Updater(spec=spec, pk_value=purger.pk_value)
```

Comment:
This seems to be completely wrong in implementation? The situation of passing purger for soft delete itself is strange.